ksql-serde/src/main/java/io/confluent/ksql/serde/json/KsqlJsonDeserializer.java

```diff
@@ -82,55 +84,53 @@ public KsqlJsonDeserializer(
   }
 
   @Override
-  public void configure(final Map<String, ?> map, final boolean b) {
+  public void configure(final Map<String, ?> map, final boolean isKey) {
+    this.target = isKey ? "key" : "value";
   }
 
   @Override
   public Object deserialize(final String topic, final byte[] bytes) {
     try {
-      final Object value = deserialize(bytes);
+      final JsonNode value = bytes == null
+          ? null
+          : MAPPER.readTree(bytes);
+
+      final Object coerced = enforceFieldType(
+          "$",
+          new JsonValueContext(value, physicalSchema.serializedSchema())
+      );
+
       if (LOG.isTraceEnabled()) {
-        LOG.trace("Deserialized value. topic:{}, row:{}", topic, value);
+        LOG.trace("Deserialized {}. topic:{}, row:{}", target, topic, coerced);
       }
-      return value;
+
+      return coerced;
     } catch (final Exception e) {
       throw new SerializationException(
-          "Error deserializing JSON message from topic: " + topic, e);
-    }
-  }
-
-  private Object deserialize(final byte[] bytes) {
-    try {
-      if (bytes == null) {
-        return null;
-      }
-
-      final JsonNode value = MAPPER.readTree(bytes);
-      return enforceFieldType(this, physicalSchema.serializedSchema(), value);
-    } catch (final IOException e) {
-      throw new SerializationException(e);
+          "mvn " + target + " from topic: " + topic, e);
     }
   }
 
   private static Object enforceFieldType(
-      final KsqlJsonDeserializer deserializer,
-      final Schema schema,
-      final JsonNode columnVal
+      final String pathPart,
+      final JsonValueContext context
   ) {
-    return enforceFieldType(new JsonValueContext(deserializer, schema, columnVal));
-  }
-
-  private static Object enforceFieldType(final JsonValueContext context) {
     if (context.val == null || context.val instanceof NullNode) {
       return null;
     }
 
-    final Function<JsonValueContext, Object> handler = HANDLERS.getOrDefault(
-        context.schema.type(),
-        type -> {
-          throw new KsqlException("Type is not supported: " + type);
-        });
-    return handler.apply(context);
+    try {
+      final Function<JsonValueContext, Object> handler = HANDLERS.getOrDefault(
+          context.schema.type(),
+          type -> {
+            throw new KsqlException("Type is not supported: " + type);
+          });
+      return handler.apply(context);
+    } catch (final CoercionException e) {
+      throw new CoercionException(e.getRawMessage(), pathPart + e.getPath(), e);
```

Comment:
Very cool! I spent some time thinking about how to implement this and didn't come up with anything nearly as nice. Glad to have had the opportunity to learn from you :)

Reply:
:)